File: backend/src/main/java/com/example/germanlearningapp/entity/GrammarQuestion.java
package com.example.germanlearningapp.entity;

import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import java.util.ArrayList;
import java.util.List;

@Entity
public class GrammarQuestion {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.EAGER, optional = false)
    @JoinColumn(name = "topic_id")
    private GrammarTopic topic;

    @Column(nullable = false, length = 500)
    private String question;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "grammar_question_options", joinColumns = @JoinColumn(name = "grammar_question_id"))
    @Column(name = "option_text", nullable = false)
    private List<String> options = new ArrayList<>();

    @Column(nullable = false)
    private Integer correctAnswerIndex;

    @Column(nullable = false, length = 1000)
    private String explanation;

    @Column(nullable = false)
    private String level;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public GrammarTopic getTopic() {
        return topic;
    }

    public void setTopic(GrammarTopic topic) {
        this.topic = topic;
    }

    public String getQuestion() {
        return question;
    }

    public void setQuestion(String question) {
        this.question = question;
    }

    public List<String> getOptions() {
        return options;
    }

    public void setOptions(List<String> options) {
        this.options = options;
    }

    public Integer getCorrectAnswerIndex() {
        return correctAnswerIndex;
    }

    public void setCorrectAnswerIndex(Integer correctAnswerIndex) {
        this.correctAnswerIndex = correctAnswerIndex;
    }

    public String getExplanation() {
        return explanation;
    }

    public void setExplanation(String explanation) {
        this.explanation = explanation;
    }

    public String getLevel() {
        return level;
    }

    public void setLevel(String level) {
        this.level = level;
    }
}

File: frontend/lib/models/grammar_question.dart
class GrammarQuestion {
  const GrammarQuestion({
    required this.id,
    required this.topicId,
    required this.topicTitle,
    required this.question,
    required this.options,
    required this.correctAnswerIndex,
    required this.explanation,
    required this.level,
  });

  final int id;
  final int topicId;
  final String topicTitle;
  final String question;
  final List<String> options;
  final int correctAnswerIndex;
  final String explanation;
  final String level;

  factory GrammarQuestion.fromJson(Map<String, dynamic> json) {
    final options = (json['options'] as List<dynamic>)
        .map((option) => option as String)
        .toList();

    return GrammarQuestion(
      id: json['id'] as int,
      topicId: json['topicId'] as int,
      topicTitle: json['topicTitle'] as String,
      question: json['question'] as String,
      options: options,
      correctAnswerIndex: json['correctAnswerIndex'] as int,
      explanation: json['explanation'] as String,
      level: json['level'] as String,
    );
  }
}
